File: app/src/test/java/com/metahumanz/pacilread/importer/BookDuplicateDetectorTest.java
package com.metahumanz.pacilread.importer;

import static org.junit.Assert.assertEquals;

import org.junit.Test;

import java.util.Arrays;
import java.util.Collections;
import java.util.Map;

public class BookDuplicateDetectorTest {
    @Test
    public void exactHashTakesPriorityOverMetadata() {
        BookDuplicateDetector.Candidate existing = new BookDuplicateDetector.Candidate(
                "old", "旧标题", "甲", "abc");
        BookDuplicateDetector.Candidate incoming = new BookDuplicateDetector.Candidate(
                "new", "新标题", "乙", "ABC");

        Map<String, BookDuplicateDetector.MatchType> result = BookDuplicateDetector.detect(
                Collections.singletonList(existing), Collections.singletonList(incoming));

        assertEquals(BookDuplicateDetector.MatchType.EXACT_CONTENT, result.get("new"));
    }

    @Test
    public void detectsMetadataAndDuplicatesInsideIncomingBatch() {
        BookDuplicateDetector.Candidate first = new BookDuplicateDetector.Candidate(
                "first", "同一本书", "作者", "one");
        BookDuplicateDetector.Candidate second = new BookDuplicateDetector.Candidate(
                "second", " 同一本书 ", "作者", "two");
        BookDuplicateDetector.Candidate third = new BookDuplicateDetector.Candidate(
                "third", "另一标题", "其他", "one");

        Map<String, BookDuplicateDetector.MatchType> result = BookDuplicateDetector.detect(
                Collections.emptyList(), Arrays.asList(first, second, third));

        assertEquals(BookDuplicateDetector.MatchType.SAME_TITLE_AUTHOR, result.get("second"));
        assertEquals(BookDuplicateDetector.MatchType.EXACT_CONTENT, result.get("third"));
    }
}

File: app/src/test/kotlin/com/metahumanz/pacilread/importer/BookDuplicateDetectorTest.kt
package com.metahumanz.pacilread.importer

import org.junit.Assert.assertEquals
import org.junit.Test

class BookDuplicateDetectorTest {
    @Test
    fun exactHashTakesPriorityOverMetadata() {
        val existing = BookDuplicateDetector.Candidate("old", "旧标题", "甲", "abc")
        val incoming = BookDuplicateDetector.Candidate("new", "新标题", "乙", "ABC")
        val result = BookDuplicateDetector.detect(listOf(existing), listOf(incoming))
        assertEquals(BookDuplicateDetector.MatchType.EXACT_CONTENT, result["new"])
    }

    @Test
    fun detectsMetadataAndDuplicatesInsideIncomingBatch() {
        val first = BookDuplicateDetector.Candidate("first", "同一本书", "作者", "one")
        val second = BookDuplicateDetector.Candidate("second", " 同一本书 ", "作者", "two")
        val third = BookDuplicateDetector.Candidate("third", "另一标题", "其他", "one")
        val result = BookDuplicateDetector.detect(emptyList(), listOf(first, second, third))
        assertEquals(BookDuplicateDetector.MatchType.SAME_TITLE_AUTHOR, result["second"])
        assertEquals(BookDuplicateDetector.MatchType.EXACT_CONTENT, result["third"])
    }
}
